src/backend/distributed/planner/deparse_shard_query.c

```diff
@@ -139,30 +141,65 @@ UpdateTaskQueryString(Query *query, Oid distributedTableId, RangeTblEntry *value
 		valuesRTE->values_lists = task->rowValuesLists;
 	}
 
-	/*
-	 * For INSERT queries, we only have one relation to update, so we can
-	 * use deparse_shard_query(). For UPDATE and DELETE queries, we may have
-	 * subqueries and joins, so we use relation shard list to update shard
-	 * names and call pg_get_query_def() directly.
-	 */
+	bool localExecution = TaskAccessesLocalNode(task);
+
 	if (query->commandType == CMD_INSERT)
 	{
-		deparse_shard_query(query, distributedTableId, task->anchorShardId, queryString);
+		/*
+		 * For INSERT queries, we only have one relation to update, so we can
+		 * use deparse_shard_query().
+		 */
+		if (localExecution)
+		{
+			/*
+			 * not all insert queries are copied before calling this
+			 * function, so we do it here
+			 */
+			query = copyObject(query);
+
+			/*
+			 * We store this in the task so we can lazily call
+			 * deparse_shard_query when the string is needed
+			 */
+			task->distributedTableId = distributedTableId;
+		}
+		else
+		{
+			deparse_shard_query(query, distributedTableId, task->anchorShardId,
+								queryString);
+		}
 	}
 	else
 	{
+		/*
+		 * For UPDATE and DELETE queries, we may have subqueries and joins, so
+		 * we use relation shard list to update shard names and call
+		 * pg_get_query_def() directly.
+		 */
 		List *relationShardList = task->relationShardList;
 		UpdateRelationToShardNames((Node *) query, relationShardList);
-
-		pg_get_query_def(query, queryString);
+		if (!localExecution)
+		{
+			pg_get_query_def(query, queryString);
+		}
 	}
```

Comment:
This function is a little dizzying. Would it be worth writing a specialized one at least for inserts?

Reply:
Simplified it now. Is this better?

Reply:
I have minimised my changes to this function and I think it's as easy to understand as it was before now.